webrender/src/device/gfx/device.rs

```diff
@@ -2076,6 +2184,60 @@ impl<B: hal::Backend> Device<B> {
         }
     }
 
+    pub fn copy_cache_buffer(&mut self, dst: &Texture, src: &Texture) {
+        assert!(src.is_buffer && dst.is_buffer);
+        let ref src_buffer = self.gpu_cache_buffers[&src.id];
+        let ref dst_buffer = self.gpu_cache_buffers[&dst.id];
+        unsafe {
+            let cmd_buffer = self.command_pool[self.next_id].command_buffer_mut();
+
+            let barriers = src_buffer
+                .transit(hal::buffer::Access::TRANSFER_READ)
+                .into_iter()
+                .chain(dst_buffer.transit(hal::buffer::Access::TRANSFER_WRITE));
+
+            cmd_buffer.pipeline_barrier(
+                hal::pso::PipelineStage::FRAGMENT_SHADER | hal::pso::PipelineStage::VERTEX_SHADER
+                    .. hal::pso::PipelineStage::TRANSFER,
+                hal::memory::Dependencies::empty(),
+                barriers,
+            );
+
+            cmd_buffer.copy_buffer(
+                &src_buffer.buffer,
+                &dst_buffer.buffer,
+                &[hal::command::BufferCopy {
+                    src: 0,
+                    dst: 0,
+                    size: src_buffer.size,
+                }],
+            );
+            if let Some(barrier) = dst_buffer.transit(hal::buffer::Access::HOST_WRITE) {
+                cmd_buffer.pipeline_barrier(
+                    hal::pso::PipelineStage::TRANSFER
+                        .. hal::pso::PipelineStage::HOST,
+                    hal::memory::Dependencies::empty(),
+                    &[barrier],
+                );
+            }
+        }
+    }
+
+    pub fn prepare_cache_buffer_for_read(&mut self) {
+        unsafe {
+            let ref buffer = &self.gpu_cache_buffers[&self.bound_gpu_cache];
+            if let Some(barrier) = buffer.transit(hal::buffer::Access::SHADER_READ) {
```

Comment:
I think that's our problem (or at least one of them :) ). GPU cache is supposed to be modified on CPU while it's read on GPU. So we can *not* just have a single `SHADER_READ` access on it here. While the command buffer is executing, our CPU will still want to write some data for the next frames. We basically need to:
  - always use SHADER_READ + HOST_WRITE access on it
  - except for resizes, where we transition it to `TRANSFER_SRC` before we copy it and then throw it out (after the copy is done on GPU)

Oh, wait, I just realized we can't copy it on GPU on resize because we would have to wait (or do something smart) before writing it on CPU afterwards, duh. Let me write something up as a separate comment...